pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1617 @@
+					),
+				})
+			}
+		}
+
+		if svcInfo.SessionAffinityType() == v1.ServiceAffinityClientIP {
+			// Generate the per-endpoint affinity sets
+			for _, ep := range allLocallyReachableEndpoints {
+				epInfo, ok := ep.(*endpointInfo)
+				if !ok {
+					klog.ErrorS(nil, "Failed to cast endpointsInfo", "endpointsInfo", ep)
+					continue
+				}
+
+				// Create a set to store current affinity mappings. The
+				// nft docs say "dynamic" is only needed for sets
+				// containing stateful objects (eg counters), but (at
+				// least on RHEL8) if we create the set without "dynamic",
+				// it later gets mutated to have it, and then the next
+				// attempt to tx.Add() it here fails because it looks like
+				// we're trying to change the flags.
+				tx.Add(&nftables.Set{
+					Name: epInfo.affinitySetName,
+					Type: ipvX_addr,
+					Flags: []nftables.SetFlag{
+						nftables.DynamicFlag,
+						nftables.TimeoutFlag,
+					},
+					Timeout: ptr.To(time.Duration(svcInfo.StickyMaxAgeSeconds()) * time.Second),
+				})
+				activeAffinitySets.Insert(epInfo.affinitySetName)
+			}
+		}
+
+		// If Cluster policy is in use, create the chain and create rules jumping
+		// from clusterPolicyChain to the clusterEndpoints
+		if usesClusterPolicyChain {
+			proxier.writeServiceToEndpointRules(tx, svcPortNameString, svcInfo, clusterPolicyChain, clusterEndpoints)
+		}
+
+		// If Local policy is in use, create rules jumping from localPolicyChain
+		// to the localEndpoints
+		if usesLocalPolicyChain {
+			proxier.writeServiceToEndpointRules(tx, svcPortNameString, svcInfo, localPolicyChain, localEndpoints)
+		}
+
+		// Generate the per-endpoint chains
+		for _, ep := range allLocallyReachableEndpoints {
+			epInfo, ok := ep.(*endpointInfo)
+			if !ok {
+				klog.ErrorS(nil, "Failed to cast endpointInfo", "endpointInfo", ep)
+				continue
+			}
+
+			endpointChain := epInfo.chainName
+
+			// Handle traffic that loops back to the originator with SNAT.
+			tx.Add(&nftables.Rule{
+				Chain: endpointChain,
+				Rule: nftables.Concat(
+					ipX, "saddr", epInfo.IP(),
+					"jump", kubeMarkMasqChain,
+				),
+			})
+
+			// Handle session affinity
+			if svcInfo.SessionAffinityType() == v1.ServiceAffinityClientIP {
+				tx.Add(&nftables.Rule{
+					Chain: endpointChain,
+					Rule: nftables.Concat(
+						"update", "@"+epInfo.affinitySetName,
+						"{", ipX, "saddr", "}",
+					),
+				})
+			}
+
+			// DNAT to final destination.
+			tx.Add(&nftables.Rule{
+				Chain: endpointChain,
+				Rule: nftables.Concat(
+					"meta l4proto", protocol,
+					"dnat to", epInfo.Endpoint,
+				),
+			})
+		}
+	}
+
+	// Figure out which chains are now stale. Unfortunately, we can't delete them
+	// right away because if there are map entries pointing to a chain, the chain
+	// doesn't become deletable until a short amount of time *after* the map entry
+	// is deleted. So we flush them now, and record that they need to be deleted
+	// in the next sync.
```

Comment:
FTR this still isn't quite right; if the BoundedFrequencyRunner allows a "burst" (or the admin sets min-sync-period to 0) then the next sync might be too soon afterward and the deletes will still fail. (This is visible in the latest kind outputs, but I don't think it's why the kind IPv6 job failed.) I'm probably going to have to add an explicit timer or something.

(Also FTR the underlying bug is fixed in very recent kernels though we wouldn't be able to depend on that for a long time...)